src/sentry/static/sentry/app/views/settings/components/forms/tableField.tsx

```diff
@@ -0,0 +1,172 @@
+    columnKeys: PropTypes.arrayOf(PropTypes.string),
+    /**
+     * Automatically save even if fields are empty
+     */
+    allowEmpty: PropTypes.bool,
+  };
+
+  static defaultProps = {
+    addButtonText: t('Add Item'),
+    perItemMapping: false,
+    allowEmpty: false,
+    // Since we're saving an object, there isn't a great way to render the
+    // change within the toast. Just turn off displaying the from/to portion of
+    // the message.
+    formatMessageValue: false,
+  };
+
+  hasValue = value => defined(value) && !objectIsEmpty(value);
+
+  renderField = props => {
+    const {
+      onChange,
+      onBlur,
+      addButtonText,
+      columnLabels,
+      columnKeys,
+      disabled,
+      allowEmpty,
+    } = props;
+
+    const mappedKeys = columnKeys;
+    const emptyValue = mappedKeys.reduce((a, v) => ({...a, [v]: null}), {id: ''});
+
+    const valueIsEmpty = this.hasValue(props.value);
+    const value = valueIsEmpty ? props.value : [];
+
+    const saveChanges = (nextValue: object) => {
+      onChange(nextValue, []);
+
+      const validValues = !Object.values(nextValue)
+        .map(o => Object.values(o).find(v => v === null))
+        .includes(undefined);
+
+      if (allowEmpty || validValues) {
+        onBlur();
+      }
+    };
+
+    const addRow = () => {
+      saveChanges([...value, emptyValue]);
+    };
+
+    const removeRow = rowIndex => {
+      //eslint-disable-next-line no-unused-vars
+      const newValue = [...value];
+      newValue.splice(rowIndex, 1);
+      saveChanges(newValue);
+    };
+
+    const setValue = (rowIndex: number, fieldKey: string, fieldValue: any) => {
+      const newValue = [...value];
+      newValue[rowIndex][fieldKey] = fieldValue.currentTarget.value;
+      saveChanges(newValue);
+    };
+
+    const button = (
+      <Button icon="icon-circle-add" onClick={addRow} size="xsmall" disabled={disabled}>
+        {addButtonText}
+      </Button>
+    );
+
+    // The field will be set to inline when there is no value set for the
+    // field, just show the button.
+    if (!valueIsEmpty) {
+      return <Box>{button}</Box>;
+    }
+
+    return (
+      <React.Fragment>
+        <Flex align="center">
+          {mappedKeys.map((fieldKey, i) => (
+            <Flex
+              key={fieldKey}
+              ml={1}
+              flex="1 0 0"
+              align="center"
+              justify="space-between"
+            >
+              <StyledHeader>{columnLabels[fieldKey]}</StyledHeader>
+              {i === mappedKeys.length - 1 && button}
+            </Flex>
+          ))}
+        </Flex>
+        {value.map((row, rowIndex) => (
+          <Flex key={rowIndex} align="center" mt={1}>
+            {mappedKeys.map((fieldKey: string, i: number) => (
+              <Flex key={fieldKey} align="center" ml={1} flex="1 0 0">
+                <Box flex={1}>
+                  <Input
+                    onChange={(v: string) => setValue(rowIndex, fieldKey, v ? v : null)}
```

Comment:
@MeredithAnya I am confused about this:
1. Why do we need the ternary statement `v ? v : null`?
2. I thought `v` has the property `currentTarget`? How could it be a string here?